### PR TITLE
[GWC-1055] Out of memory due to premature rendering surface creation

### PR DESCRIPTION
**Pull request by turingtestfail**
added test

cleanup

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]

File: geowebcache/wms/src/main/java/org/geowebcache/service/wms/BufferedImageWrapper.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * @author Joseph Miller, Copyright 2022
+ */
+package org.geowebcache.service.wms;
+
+import java.awt.Color;
+import java.awt.Graphics2D;
+import java.awt.RenderingHints;
+import java.awt.image.BufferedImage;
+
+public class BufferedImageWrapper {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/main/java/org/geowebcache/service/wms/BufferedImageWrapper.java]: fixed

File: geowebcache/wms/src/main/java/org/geowebcache/service/wms/WMSTileFuser.java [hunk 2]
```diff
@@ -703,19 +688,26 @@ protected void renderCanvas()
...
                                 + Arrays.toString(gridLoc));
-
-                gfx.drawImage(tileImg, canvasx, canvasy, null); // imageObserver
+                // creating canvase only after first tile is acquired in order to avoid premature
+                // canvas creation
+                // in the case of locks
+                if (bufferedImageWrapper == null) {
+                    createCanvas();
+                }
+                bufferedImageWrapper.drawImage(tileImg, canvasx, canvasy); // imageObserver
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/main/java/org/geowebcache/service/wms/WMSTileFuser.java]: fixed

File: geowebcache/wms/src/main/java/org/geowebcache/service/wms/WMSTileFuser.java [hunk 3]
```diff
@@ -703,19 +688,26 @@ protected void renderCanvas()
                                 + canvasy
                                 + ",null) "
                                 + Arrays.toString(gridLoc));
-
-                gfx.drawImage(tileImg, canvasx, canvasy, null); // imageObserver
+                // creating canvase only after first tile is acquired in order to avoid premature
+                // canvas creation
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/main/java/org/geowebcache/service/wms/WMSTileFuser.java]: fixed

File: geowebcache/wms/src/main/java/org/geowebcache/service/wms/WMSTileFuser.java [hunk 4]
```diff
@@ -703,19 +688,26 @@ protected void renderCanvas()
...
+                // in the case of locks
+                if (bufferedImageWrapper == null) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/main/java/org/geowebcache/service/wms/WMSTileFuser.java]: fixed

File: geowebcache/wms/src/main/java/org/geowebcache/service/wms/BufferedImageWrapper.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            gfx = (Graphics2D) getCanvas().getGraphics();
+        }
+        if (bgColor != null) {
+            gfx.setColor(bgColor);
+            gfx.fillRect(0, 0, canvasSize[0], canvasSize[1]);
+        }
+        return gfx;
+    }
+
+    public void setGfx(Graphics2D gfx) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/main/java/org/geowebcache/service/wms/BufferedImageWrapper.java]: fixed

File: geowebcache/wms/src/test/java/org/geowebcache/service/wms/WMSTileFuserTest.java [hunk 6]
```diff
@@ -120,6 +125,63 @@ public void testTileFuserResolution() throws Exception {
...
+        Mockito.when(sb.get(ArgumentMatchers.any(TileObject.class)))
+                .thenAnswer(
+                        invoc -> {
+                            TileObject stObj = (TileObject) invoc.getArguments()[0];
+                            final File imageTile =
+                                    new File(getClass().getResource("/image.png").toURI());
+                            stObj.setBlob(new FileResource(imageTile));
+                            stObj.setCreated((new Date()).getTime());
+                            stObj.setBlobSize(1000);
+                            if (count.getAndIncrement() == 0) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/test/java/org/geowebcache/service/wms/WMSTileFuserTest.java]: fixed

File: geowebcache/wms/src/test/java/org/geowebcache/service/wms/WMSTileFuserTest.java [hunk 7]
```diff
@@ -120,6 +125,63 @@ public void testTileFuserResolution() throws Exception {
...
+                                assertNull(tileFuser.bufferedImageWrapper);
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/test/java/org/geowebcache/service/wms/WMSTileFuserTest.java]: fixed

File: geowebcache/wms/src/main/java/org/geowebcache/service/wms/BufferedImageWrapper.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+            if (hints != null) {
+                hintsTemp = hints;
+            }
+            gfx.addRenderingHints(hintsTemp);
+        }
+        return canvas;
+    }
+
+    public void setCanvas(BufferedImage canvas) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply geowebcache/wms/src/main/java/org/geowebcache/service/wms/BufferedImageWrapper.java]: fixed

File: geowebcache/wms/src/test/java/org/geowebcache/service/wms/WMSTileFuserTest.java [hunk 9]
```diff
@@ -120,6 +125,60 @@ public void testTileFuserResolution() throws Exception {
...
+                                    new File(getClass().getResource("/image.png").toURI());
+                            stObj.setBlob(new FileResource(imageTile));
+                            stObj.setCreated((new Date()).getTime());
+                            stObj.setBlobSize(1000);
+                            if (count.incrementAndGet() == 1) {
+                                // first time, canvas should be null
+                                assertNull(tileFuser.bufferedImageWrapper.canvas);
+                                return true;
+                            } else {
+                                count.incrementAndGet();
```
Comment (aaime) [inline geowebcache/wms/src/test/java/org/geowebcache/service/wms/WMSTileFuserTest.java:161]: Double increment here: the first increment happened in the if, it happens whether the condition is matched, or not.

**Comment by aaime** [comment]
Thanks! Can you backport to 2.21.x as well?

**Pull request merged**